monitoring/mongodb/alerts.yaml

```diff
@@ -139,3 +139,45 @@ groups:
     annotations:
       description: 'MongoDb has low disk space'
       summary: 'MongoDb has low disk space'
+  
+  - alert: MongoDbNodeRecovering
+    expr: |
+      count_over_time(mongodb_rs_members_state{namespace="${namespace}", pod=~"${service}.*", rs_state="3"}[1h]) > 0
+    for: 1h
+    labels:
+      severity: warning
+    annotations:
+      description: "Mongodb pod `{{ $labels.pod }}` has been in the 'RECOVERING' state at least once in the past hour. The instance may not be able to join the replica set if the platform ingests a large number of operations during this time. This alert is expected if the 'Resync a Data Services MongoDB Member' procedure has recently been executed."
+      summary: MongoDB node recovering for too long
+
+  - alert: MongoDbInvalidState
+    expr: |
+      avg_over_time(mongodb_rs_members_state{namespace="${namespace}", pod=~"${service}.*", rs_state=~"6|8|10"}[1m]) > 0
+    for: 5m
+    labels:
+      severity: critical
+    annotations:
+      description: "Mongodb pod `{{ $labels.pod }}` is in the `{{ $labels.rs_state }}` state. Immediate attention is required."
+      summary: "MongoDB node in an invalid state: 6 (UNKNOWN), 8 (DOWN), 10 (REMOVED)"
+
+
+  - alert: MongoDbNodeStartup2
+    expr: |
+      avg_over_time(mongodb_rs_members_state{namespace="${namespace}", pod=~"${service}.*", rs_state="5"}[1h]) == 5
+    for: 1h
+    labels:
+      severity: warning
+    annotations:
+      description: "Mongodb pod `{{ $labels.pod }}` is in the 'STARTUP2' state for an hour. The instance might be stuck."
+      summary: MongoDB node in STARTUP2 state for too long
+
+  - alert: MongoDbNodeNotSynced
+    expr: |
+      sum by (rs_nm) (mongodb_rs_members_state{namespace="${namespace}", pod=~"${service}.*", member_state="SECONDARY"}) != (${replicas} - 1)
+    for: 1m
+    labels:
+      severity: critical
+    annotations:
+      description: "Mongodb replica set `{{ $labels.rs_nm }}` is out of sync. One or more pods are not receiving data and must be added back to the cluster to avoid performance and storage problems."
+      summary: MongoDb replica set out of sync
```

Comment:
Won't this alert always fire if we have one replica that is recovering/other? Here, we fire the alert if during the last minute, we do not see e.g. 2 SECONDAY on a 3-node platform. However, it is totally acceptable to have STARTUP2 or RECOVERING periodically (during high loads, for example). We already detect these cases in the above alerts.

I think what we want with this alert is a longer check duration (it is less sensitive than the primary one, that blocks everything). And then we explain in the description that the instance is missing N MongoDB secondaries, meaning that in case of primary loss we might have metadata loss.

What remains is the duplication of alerts; if one pod is RECOVERING for let's say 5min, we will have both this alert, and later, the MongoDbNodeRecovering one...

Reply:
Indeed , but I think that if we increase the evaluation duration to 1h we will have the first warning ( in case of a recovery or other state ) then we will have the critical `MongoDbNodeNotSynced` , IMHO keeping both is okay what do you think ? cc : @francoisferrand

Reply:
I think 10m is acceptable, as long as we are alerted before the oplog has any risk of being overriden, we are safe (1h can be a bit too long for some platforms today...)
A mongodb instance that is recovering due to usual operations won't stay in this state for too long.